examples/add_points_with_multi_color_text.py

```diff
@@ -0,0 +1,50 @@
+"""
+Display a points layer on top of an image layer using the add_points and
+add_image APIs
+"""
+
+import numpy as np
+import napari
+from napari.layers.utils.property_map import PropertyMap
+
+# add the image
+viewer = napari.view_image(np.zeros((400, 400)))
+# add the points
+points = np.array([[100, 100], [200, 300], [333, 111]])
+
+# create properties for each point
+properties = {
+    'confidence': np.array([1, 0.5, 0]),
+    'good_point': np.array([True, False, False]),
+}
+
+good_point_color = {
+    False: 'green',
+    True: 'blue',
+}
+
+text = {
+    'text': 'Confidence is {confidence:.2f}',
+    'size': 20,
+    'color': PropertyMap.from_discrete_map('good_point', good_point_color),
```

Comment:
Could I achieve this without using `from_discrete_map`. Maybe instead pass a dict or a tuple to `color` which had `map: 'discrete'` or something like ('good_point', good_point_color, 'discrete'). Just thinking about serialization etc

Reply:
Yeah, I also dislike this line, though not so much for serialization... this line feels un-pythonic to me. (Whatever that means! 😂) What I'd probably like to write, as a user, is something like:

```python
text = {
    'text': 'Confidence is {confidence:.2f}',
    'size': 20,
    'color': {
        'source': 'good_point',  # or 'property': 'good_point'?
        'map': good_point_color,  # or good_point_color.get
    },
    'translation': [-30, 0],
}
```

The dictionary keys would map to attribute names in PropertyMap.

Unfortunately @sofroniewn, some maps are functions and thus not serializable — I think?

Reply:
will take a deeper look soon, but my first thought here is that there's a nice feature of pydantic using Union (described [here](https://pydantic-docs.helpmanual.io/usage/types/#unions)) that could let you avoid all of these classmethods, and glean the intended type automatically based on the input dict like @jni is suggesting.

Consider the following model:
```python
from typing import TypeVar, Generic, List, Union
from typing_extensions import Literal
from pydantic import BaseModel
from napari.utils import Colormap

T = TypeVar("T")


class _PropertyMap(BaseModel, Generic[T]):
    values: List[T] = []


class DirectPropertyMap(_PropertyMap[T]):
    default_value: T
    constant: bool = False


class ConstantPropertyMap(DirectPropertyMap[T]):
    constant: Literal[True] = True


class NamedPropertyMap(_PropertyMap[T]):
    name: str


class NamedPropertyDiscreteMap(NamedPropertyMap[T]):
    discrete_map: dict[str, T]


class NamedPropertyColorMap(NamedPropertyMap[T]):
    colormap: Colormap

# note, these should be ordered from most specific to least specific
PropertyMap = Union[
    NamedPropertyColorMap,
    NamedPropertyDiscreteMap,
    NamedPropertyMap,
    DirectPropertyMap,
    ConstantPropertyMap,
]

class Model(BaseModel):
    prop_map: PropertyMap
```

You can get all of these subclasses based on the input alone.  Note in the example below how pydantic gives you the correct model instance based on the input.  Sometimes this requires a little clever type narrowing (using, e.g. `Literal[True]` for the "Constant" type).

```python
In [5]: Model(prop_map={})
Out[5]: Model(prop_map=DirectPropertyMap(values=[], default_value=None, constant=False))

In [6]: Model(prop_map={'default_value': 1})
Out[6]: Model(prop_map=DirectPropertyMap(values=[], default_value=1, constant=False))

In [7]: Model(prop_map={'default_value': 1, 'constant': True})
Out[7]: Model(prop_map=ConstantPropertyMap(values=[], default_value=1, constant=True))

In [8]: Model(prop_map={'name': 'somename'})
Out[8]: Model(prop_map=NamedPropertyMap(values=[], name='somename'))

In [9]: Model(prop_map={'name': 'somename', 'discrete_map': {'key': 'val'}})
Out[9]: Model(prop_map=NamedPropertyDiscreteMap(values=[], name='somename', discrete_map={'key': 'val'}))

In [10]: Model(prop_map={'name': 'somename', 'colormap': {'colors': [0,1,1,1]}})
Out[10]: Model(prop_map=NamedPropertyColorMap(values=[], name='somename', colormap=Colormap(colors=array([[0., 1., 1., 1.]]), name='custom', interpolation=<ColormapInterpolationMode.LINEAR: 'linear'>, controls=array([0.]))))
```

Reply:
I went with factory methods here for a few reasons.
- Only one import is needed (`PropertyMap`) to create any kind of style attribute (color, size, etc.).
- The color mapping type is explicitly defined by the method name.
- The color mappings returned by these factory methods are serializable (they are all `EventedModels`).

I was thinking about adding support for `dict`, from which we could infer the type of mapping. The major upside of this is that no imports are needed and the user never needs to know about `PropertyMap`. The downsides are that we need to add code to handle the type inference, and we need to explain the logic/precedence of that inference to the user. Let me try that out in this PR to see how easy that is to implement.

One issue with this approach is the ambiguity of something like `'color': {'source': 'good_point'}` or maybe even `'color': 'good_point'`. If I understand correctly, the way that Altair/Vega handles that is by looking up the data type of the property/data and using some default/sensible discrete/continuous color map depending on the type (e.g. quantitative, nominal). 

napari does this for some types, but not for others - e.g. when the property is quantitative/continuous it uses the default colormap, but when it's discrete/nominal it defaults to white.

Reply:
> The downsides are that we need to add code to handle the type inference, and we need to explain the logic/precedence of that inference to the user.

doesn't my example above suggest that you can _remove_ a lot of code, and get the type inference for free?  With no imports at all?

Reply:
@tlambert03 : yep, sorry, that reply was mostly written before your response. The boilerplate code removal looks nice, though the need for `Literal` (though I'm not sure it's actually needed because `ConstantPropertyMap` doesn't need a `default_value`) and sensitivity to ordering is a little concerning.

I think my bigger concern here is around explaining the interface to the user. If we expect users to pass `dict`, then we still have to point them to the class definition/docs to explain what the valid keys/values of that dict are.

Reply:
> though I'm not sure it's actually needed because ConstantPropertyMap doesn't need a default_value

I tried without it, but it looks like if there are two models that declare `any_name: Any`, then it will go with the first one, even if the attribute name doesn't match exactly.  Feel free to play with it, but that's what I had to do to get it to work.  (and I think it's reasonable to think of `ConstantProperty` as something with a `default_value` that doesn't change)

> sensitivity to ordering is a little concerning.

it's very testable... so that part doesn't concern me...

> explaining the interface to the user. If we expect users to pass dict, then we still have to point them to the class definition/docs to explain what the valid keys/values of that dict are.

yeah, I can't think of a way to completely circumvent needing to explain something to the user if they're trying to indicate something special about a property... but it doesn't feel like we're having to teach them anything more to use this:

```py
text = {
    ...
    'color': {'name': 'good_point', 'discrete_map': good_point_color},
    ...
}
```
than this:

```py
text = {
    ...
    'color': PropertyMap.from_discrete_map('good_point', good_point_color),
    ...
}
```

... right?  i.e. in both cases, they need to understand that we have a "discrete map" offering, and you can get it in case A) by using the `discrete_map` key in your dict, and in B) by importing `PropertyMap` and using it's `from_discrete_map` method?

Reply:
Agreed ultimately on the doc problem. One possible benefit of forcing the user to import `PropertyMap`, there's a direct/explicit connection between the call they have to make and the docs (i.e. they can do something like `help(PropertyMap)`, `help(text.color)` or `help(PropertyMap.from_discrete_map)`), but that's not a big win.

The problem with the inability to distinguish between direct/constant without the need for a `Literal` gets worse when we consider the string maps I defined so far because all 4 have the same field/type signature. I think you could question if those 4 are really needed, but it suggests a design problem that might bite us in the future.

I found that pydantic's `GenericModel` gives us one way to distinguish between them. Consider a similar example to the one above:

```python
from typing import TypeVar, Generic, List, Union
from pydantic import BaseModel
from pydantic.generics import GenericModel

T = TypeVar("T")


class _PropertyMap(Generic[T]):
    pass


class DirectPropertyMap(GenericModel, _PropertyMap[T]):
    values: List[T] = []
    default_value: T


class ConstantPropertyMap(GenericModel, _PropertyMap[T]):
    values: List[T] = []
    constant: T


class NamedPropertyMap(GenericModel, _PropertyMap[T]):
    values: List[T] = []
    name: str


class TextFormatPropertyMap(GenericModel, _PropertyMap[str]):
    values: List[T] = []
    format_string: str


# note, these should be ordered from most specific to least specific
StringPropertyMap = Union[
    TextFormatPropertyMap,
    NamedPropertyMap[str],
    DirectPropertyMap[str],
    ConstantPropertyMap[str],
]


class Model(BaseModel):
    prop_map: StringPropertyMap


print(Model(prop_map={'default_value': 'some_default'}))
print(Model(prop_map={'constant': 'some_constant'}))
print(Model(prop_map={'name': 'some_name'}))
print(Model(prop_map={'format_string': 'some {format} {string}'}))
```

For me this gives the following output:

```
prop_map=DirectPropertyMap[str](values=[], default_value='some_default')
prop_map=ConstantPropertyMap[str](values=[], constant='some_constant')
prop_map=NamedPropertyMap[str](values=[], name='some_name')
prop_map=TextFormatPropertyMap(values=[], format_string='some {format} {string}')
```

I can't find a way to declare the `values` field in `_PropertyMap`, so it doesn't declare anything interesting here, but I also wouldn't hate that as a pure interface/protocol. Another maybe bigger problem is that this doesn't take into account events, so we might need a `GenericEventedModel`, which is not ideal.

Reply:
Actually, after re-reading the Pydantic docs, I was able to shove values into `_PropertyMap` if that's more desirable by doing

```python
...
class _PropertyMap(GenericModel, Generic[T]):
    values: List[T] = []

class DirectPropertyMap(_PropertyMap[T], Generic[T]):
    default_value: T
...
```

Reply:
> One possible benefit of forcing the user to import PropertyMap, there's a direct/explicit connection between the call they have to make and the docs (i.e. they can do something like help(PropertyMap), help(text.color) or help(PropertyMap.from_discrete_map)), but that's not a big win.

yeah, i think i'd still prefer a way not to have to import